src/renderers/webgl/WebGLBindingStates.js

```diff
@@ -0,0 +1,532 @@
+				currentState.program = program.id;
+				currentState.wireframe = wireframe;
+
+				updateBuffers = true;
+
+			}
+
+		}
+
+		if ( index !== null ) {
+
+			attributes.update( index, gl.ELEMENT_ARRAY_BUFFER );
+
+		}
+
+		if ( updateBuffers ) {
+
+			setupVertexAttributes( material, program, geometry );
+
+			if ( index !== null ) {
+
+				gl.bindBuffer( gl.ELEMENT_ARRAY_BUFFER, attributes.get( index ).buffer );
+
+			}
+
+		}
+
+	}
+
+	function createVertexArrayObject() {
+
+		if ( capabilities.isWebGL2 ) return gl.createVertexArray();
+
+		return extension.createVertexArrayOES();
+
+	}
+
+	function bindVertexArrayObject( vao ) {
+
+		if ( capabilities.isWebGL2 ) return gl.bindVertexArray( vao );
+
+		return extension.bindVertexArrayOES( vao );
+
+	}
+
+	function deleteVertexArrayObject( vao ) {
+
+		if ( capabilities.isWebGL2 ) return gl.deleteVertexArray( vao );
+
+		return extension.deleteVertexArrayOES( vao );
+
+	}
+
+	function getBindingState( geometry, program, material ) {
+
+		var wireframe = ( material.wireframe === true );
+
+		var programMap = bindingStates[ geometry.id ];
+
+		if ( programMap === undefined ) {
+
+			programMap = {};
+			bindingStates[ geometry.id ] = programMap;
+
+		}
+
+		var stateMap = programMap[ program.id ];
+
+		if ( stateMap === undefined ) {
+
+			stateMap = {};
+			programMap[ program.id ] = stateMap;
+
+		}
+
+		var state = stateMap[ wireframe ];
+
+		if ( state === undefined ) {
+
+			state = createBindingState( createVertexArrayObject() );
+			stateMap[ wireframe ] = state;
+
+		}
+
+		return state;
+
+	}
+
+	function createBindingState( vao ) {
+
+		var newAttributes = [];
+		var enabledAttributes = [];
+		var attributeDivisors = [];
+
+		for ( var i = 0; i < maxVertexAttributes; i ++ ) {
+
+			newAttributes[ i ] = 0;
+			enabledAttributes[ i ] = 0;
+			attributeDivisors[ i ] = 0;
+
+		}
+
+		return {
+
+			// for backward compatibility on non-VAO support browser
+			geometry: null,
+			program: null,
+			wireframe: false,
+
+			newAttributes: newAttributes,
+			enabledAttributes: enabledAttributes,
+			attributeDivisors: attributeDivisors,
+			object: vao,
+			attributes: {}
+
+		};
+
+	}
+
+	function needsUpdate( geometry ) {
+
+		var cachedAttributes = currentState.attributes;
+		var geometryAttributes = geometry.attributes;
+
+		if ( Object.keys( cachedAttributes ).length !== Object.keys( geometryAttributes ).length ) return true;
+
+		for ( var key in geometryAttributes ) {
+
+			var cachedAttribute = cachedAttributes[ key ];
+			var geometryAttribute = geometryAttributes[ key ];
+
+			if ( cachedAttribute.attribute !== geometryAttribute ) return true;
+
+			if ( cachedAttribute.version !== geometryAttribute.version2 ) return true;
+
+			if ( cachedAttribute.data.buffer !== geometryAttribute.data ) return true;
+
+			if ( geometryAttribute.data &&
+				cachedAttribute.data.version !== geometryAttribute.data.version2 ) return true;
+
+		}
+
+		return false;
```

Comment:
Here is the part where checking if `geometry.attributes` has an update. If we accept the API changes I wrote in PR comment, here will be much more efficient just like checking `geometry.version`.

Reply:
Doing all these checks is a bit unfortunate and feels like a hack. However, it's still better do it this way than having no VAO support at all. I hope we can make `BufferAttribute` a bit more strict in the future so parts of `needsUpdate()` won't be necessary anymore.

Reply:
Agreed with that. Added comment as a reminder.

Reply:
A bug is appeared, in our project iTowns/itowns/pull/1444, after that this PR was merged and seem to come from that function `needsUpdate`. 

The error is 
> WebGL: INVALID_OPERATION: drawElements: no buffer is bound to enabled attribute

`needsUpdate` doesn't check if the **program attributes** are identical to the **cached attributes**.  The **program attributes** can change depending on the conditional statements then the bound buffers aren't correct and it returns this error.

**fix proposal** (I'm not sure it's all right)
```js
function needsUpdate( geometry, program ) {
	const cachedAttributes = currentState.attributes;
	const geometryAttributes = geometry.attributes;
	const programAttributes = program.getAttributes();

	const cachedAttributesCount = Object.keys( cachedAttributes ).length;

	if ( cachedAttributesCount !== Object.keys( geometryAttributes ).length || 
		cachedAttributesCount !== Object.keys( programAttributes ).length
	 ) return true;

	for ( const key in cachedAttributes ) {

		const cachedAttribute = cachedAttributes[ key ];
		const geometryAttribute = geometryAttributes[ key ];
		const programAttribute = programAttributes[ key ];

		if ( cachedAttribute.attribute !== geometryAttribute ) return true;

		if ( cachedAttribute.data !== geometryAttribute.data ) return true;

		if ( programAttribute == undefined ) return true;

	}
	return false;
}
```

Reply:
Thanks for the bug report!

> The program attributes can change depending on the conditional statements

I'm investigating and found that [program attributes are initialized when `program.getAttributes()` is first called and they are cached.](https://github.com/mrdoob/three.js/blob/dev/src/renderers/webgl/WebGLProgram.js#L810-L820) So wondering how they can change. Would you tell me the condition where they can change if you already figured out? Or can you share the code snippet which can reproduce the error?

Reply:
If the material preprocessor instructions change (`#define`, `#if`, etc.) then the `WebGLProgram` instance won't be the same so the program attributes and cache attributes may be different.

Reply:
My understanding is, if shader code or preprocessor instruction changes then a new `rogram` is created or switching to another existing `program`

https://github.com/mrdoob/three.js/blob/r119/src/renderers/WebGLRenderer.js#L1311-L1353
https://github.com/mrdoob/three.js/blob/r119/src/renderers/webgl/WebGLPrograms.js#L345-L374

If a new `program` is created, a new binding state is created and `needsUpdate` returns `true` because new `cachedAttributes` of the new binding state is an empty object.

https://github.com/mrdoob/three.js/blob/r119/src/renderers/webgl/WebGLBindingStates.js#L100-L133
https://github.com/mrdoob/three.js/blob/r119/src/renderers/webgl/WebGLBindingStates.js#L166-L186

If switching to another existing `program`, an existing binding state corresponding to the `program` would be selected by `getBindingState()` and its cache should be valid to the `program`.

So I'm wondering how to reproduce the error. Could you share the code snippet which can reproduce this error?

Reply:
I found this webGL error origin :
**WebGL: INVALID_OPERATION: drawElements: no buffer is bound to enabled attribute**

It's **`geometry.dispose()`** method calling,  whose attributes are **shared** with other geometries. 

This bug occurred with **`VAOs`** in **THREE v118**. 
Even if it didn't appear in the previous versions, but it's really a bug? 

In your library, we fixed this error.  Your garbage collector avoids to dispose shared attributes.

I have created 2 JsFidlle examples that simulate asynchronous disposes by your garbage collector:

* [JsFiddle example three v117.1 without webgl error](https://jsfiddle.net/iTownsIGN/9ax61ebL/9/) 
* [JsFiddle example three v118.0 with webgl error](https://jsfiddle.net/iTownsIGN/borcfx0p/3/)

The error is still present in THREE 121.

Reply:
Thanks for the report. I confirmed the error happens.

I think this is a bug in `WebGLGeometries` (or `WebGLAttributes`), rather than in `VAO(WebGLBindingStates)`, and the bug have existed even since before we adopted `VAO`.

As you speculated, the root issue seems that `geometry.dispose()` deletes webgl buffer of shared attribute even if it's still used by other geometry.

https://github.com/mrdoob/three.js/blob/r121/src/renderers/webgl/WebGLAttributes.js#L107-L121
https://github.com/mrdoob/three.js/blob/r121/src/renderers/webgl/WebGLGeometries.js#L15-L25

You seem to have already fixed this problem in your side. So would you like to make PR?

Reply:
I avoid the problem by deleting the shared attribute properties before calling `geometry.dispose()`,
because I know that the index and position attributes are shared.

```js
geometry.index = null;
delete geometry.attributes.position;
geometry.dispose();
```
But I avoid the problem, I don't fix it.

How would you solve the problem? 

You would need to know the number of times the attribute is shared and delete it when it's 0.
This to be a property of the attribute?

Reply:
Yes, counting the reference is the solution I thought of.